reactos/drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2452 @@
+  
+        InitializeListHead(&OhciExtension->IntStaticED[ix].Link);
+  
+        StaticED += 1;
+        SchedulePA += sizeof(OHCI_ENDPOINT_DESCRIPTOR);
+    }
+
+    OhciHCCA = (POHCI_HCCA)OhciExtension->HcResourcesVA;
+    DPRINT_OHCI("OHCI_InitializeSchedule: OhciHCCA - %p\n", OhciHCCA);
+
+    for (ix = 0, jx = 31; ix < 32; ix++, jx++)
+    {
+        static UCHAR Balance[32] =
+        {0, 16, 8, 24, 4, 20, 12, 28, 2, 18, 10, 26, 6, 22, 14, 30, 
+         1, 17, 9, 25, 5, 21, 13, 29, 3, 19, 11, 27, 7, 23, 15, 31};
+  
+        OhciHCCA->InterrruptTable[Balance[ix]] =
+            (POHCI_ENDPOINT_DESCRIPTOR)(OhciExtension->IntStaticED[jx].PhysicalAddress);
+  
+        OhciExtension->IntStaticED[jx].pNextED =
+            (PULONG)&OhciHCCA->InterrruptTable[Balance[ix]];
+
+        OhciExtension->IntStaticED[jx].HccaIndex = Balance[ix];
+    }
+
+    DPRINT_OHCI("OHCI_InitializeSchedule: ix - %x\n", ix);
+
+    InitializeListHead(&OhciExtension->ControlStaticED.Link);
+
+    OhciExtension->ControlStaticED.HeadIndex = ED_EOF;
+    OhciExtension->ControlStaticED.Type = OHCI_NUMBER_OF_INTERRUPTS + 1;
+    OhciExtension->ControlStaticED.pNextED = &OperationalRegs->HcControlHeadED;
+
+    InitializeListHead(&OhciExtension->BulkStaticED.Link);
+
+    OhciExtension->BulkStaticED.HeadIndex = ED_EOF;
+    OhciExtension->BulkStaticED.Type = OHCI_NUMBER_OF_INTERRUPTS + 2;
+    OhciExtension->BulkStaticED.pNextED = &OperationalRegs->HcBulkHeadED;
+
+    FrameInterval.AsULONG = READ_REGISTER_ULONG(&OperationalRegs->HcFmInterval.AsULONG);
+
+    if ((FrameInterval.FrameInterval) < ((12000 - 1) - 120) ||
+        (FrameInterval.FrameInterval) > ((12000 - 1) + 120)) // FIXME 10%
+    {
+        FrameInterval.FrameInterval = (12000 - 1);
+    }
+
+    FrameInterval.FrameIntervalToggle = 1;
+    FrameInterval.FSLargestDataPacket = 
+        ((FrameInterval.FrameInterval - MAXIMUM_OVERHEAD) * 6) / 7;
+
+    OhciExtension->FrameInterval = FrameInterval;
+
+    DPRINT_OHCI("OHCI_StartController: FrameInterval - %lX\n",
+                FrameInterval.AsULONG);
+
+    /* reset */
+    WRITE_REGISTER_ULONG(&OperationalRegs->HcCommandStatus.AsULONG,
+                         1);
+
+    KeStallExecutionProcessor(25);
+
+    Control.AsULONG = READ_REGISTER_ULONG(&OperationalRegs->HcControl.AsULONG);
+    Control.HostControllerFunctionalState = OHCI_HC_STATE_RESET;
+
+    WRITE_REGISTER_ULONG(&OperationalRegs->HcControl.AsULONG,
+                         Control.AsULONG);
+
+    KeQuerySystemTime(&CurrentTime);
+    CurrentTime.QuadPart += 5000000; // 0.5 sec
+
+    while (TRUE)
+    {
+        WRITE_REGISTER_ULONG(&OperationalRegs->HcFmInterval.AsULONG,
+                             OhciExtension->FrameInterval.AsULONG);
+
+        FrameInterval.AsULONG =
+            READ_REGISTER_ULONG(&OperationalRegs->HcFmInterval.AsULONG);
+
+        KeQuerySystemTime(&SystemTime);
+
+        if (SystemTime.QuadPart >= CurrentTime.QuadPart)
```

Comment:
may a KeStallExecutionProcessor combined with a for loop might be a better choice?